src/4_cache_layer.js

```diff
@@ -0,0 +1,102 @@
+// Copyright (C) 2018 Bluzelle
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License, version 3,
+// as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+
+const database_pb = require('../proto/database_pb');
+const assert = require('assert');
+
+module.exports = class Cache {
+
+    constructor({onIncomingMsg, onOutgoingMsg}) {
+
+        this.onIncomingMsg = onIncomingMsg;
+        this.onOutgoingMsg = onOutgoingMsg;
+
+        this.cache = new Map();
+        this.retries = new Map();
+
+        this.MAX_RETRIES = 3;
+        this.RETRY_TIMEOUT = 300;
+
+    }
+
+
+    sendIncomingMsg(msg) {
+
+        assert(msg instanceof database_pb.database_response);
+
+        const nonce = msg.getHeader().getNonce();
+
+        assert(this.cache.has(nonce));
+        assert(this.retries.has(nonce));
+
+
+        if(msg.hasError() && msg.getError().getMessage() === 'NO CONNECTION' || 'ELECTION IN PROGRESS') {
```

Comment:
I don't think PBFT will throw a ELECTION_IN_PRGROESS error. I think Paul is working on view change integration now. Perhaps it's worth discussing this with him? AFAIK, if a primary dies PBFT the next primary in line just takes over.

Reply:
@paularchard Will the daemon ever throw an ELECTION_IN_PRGROESS error? Is there a case when it would be unavailable to commit a request?